File: p035.c
#include<stdbool.h>
#include<stdio.h>

#define EXCLUSIVE_UPPER_BOUND 1000000
#define SQRT_EXCLUSIVE_UPPER_BOUND 1000

bool is_prime[EXCLUSIVE_UPPER_BOUND];

void initialize_sieve_of_eratosthenes() {
	for(int i=2; i<EXCLUSIVE_UPPER_BOUND; i++) 
		is_prime[i] = true;
	for(int i=2; i<SQRT_EXCLUSIVE_UPPER_BOUND; i++)
		if(is_prime[i] == true)
			for(int multiple=i*i; multiple<EXCLUSIVE_UPPER_BOUND; multiple+=i)
				is_prime[multiple] = false;
}

int main() {
	initialize_sieve_of_eratosthenes();

	int count = 4; // 2, 3, 5 and 7 are circular primes but we start the loop later at 11
	for(int power_of_10 = 10; power_of_10 < EXCLUSIVE_UPPER_BOUND; power_of_10 *= 10) {
		for(int n = power_of_10 + 1; n < power_of_10 * 10; n++) {
			if(is_prime[n] == true) {
				int rotate = n;
				do {
					rotate = (rotate % 10) * power_of_10 + rotate / 10;
				} while(rotate != n && is_prime[rotate]);
				if(rotate == n)
					count++;
			}
		}
	}

	printf("%d\n", count);
}

File: p035.rs
mod prime_sieve;

fn is_circular_prime(sieve: &prime_sieve::PrimeSieve, power_of_10: usize, n: usize) -> bool {
    let mut rotation = n;
    while sieve.is_prime(rotation) {
        rotation = (rotation % 10) * power_of_10 + rotation / 10;
        if rotation == n {
            return true;
        }
    }
    false
}

fn main() {
    let sieve = prime_sieve::PrimeSieve::new(1_000_000);
    let mut count = 4; // 4 circular primes below 10 : 2, 3, 5, 7
    for power_of_10 in (1..6).map(|x| 10usize.pow(x)) {
        for n in ((power_of_10 + 1)..(power_of_10 * 10)).step_by(2) {
            if is_circular_prime(&sieve, power_of_10, n) {
                count += 1;
            }
        }
    }
    println!("{}", count);
}